### PR TITLE
Fix:Elevated limits quota on params

### PR DESCRIPTION
**Pull request by pubalokta**
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description
We're changing the way we provide elevated limits params. 
We identified providing the quota and erl_activation_period on the bucket configuration could cause an overlap between different buckets, so we decided to move those params to the `takeElevated()` call.

Now, buckets for elevated limits will be configured as:
```
{
  size: ...
  per_second:...,
  elevated_limits: {
    size: ...
    per_second:...
  }
}
```
and the takeElevated call will receive the elevated limits params on the opts object:

```
takeElevated(type, key, count, { elevated_limits: { erl_is_active_key, erl_quota_key, erl_activation_period_seconds, quota_per_calendar_month } })
```

> Describe the purpose of this PR along with any background information and the impacts of the proposed change. For the benefit of the community, please do not assume prior context.
>
> Provide details that support your chosen implementation, including: breaking changes, alternatives considered, changes to the API, etc.
>
> If the UI is being changed, please provide screenshots.

### References

> Include any links supporting this change such as a:
>
> - GitHub Issue/PR number addressed or fixed
> - Auth0 Community post
> - StackOverflow post
> - Support forum thread
> - Related pull requests/issues from other repos
>
> If there are no references, simply delete this section.

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not the default branch

### CODE REVIEW

File: test/utils.tests.js [hunk 1]
```diff
@@ -147,13 +139,106 @@ describe('utils', () => {
...
+      assert.equal(result.per_interval, bucketKeyConfig.elevated_limits.per_interval);
+      assert.equal(result.ttl, bucketKeyConfig.elevated_limits.ttl);
+      assert.equal(result.drip_interval, bucketKeyConfig.elevated_limits.drip_interval);
+      assert.equal(result.erl_activation_period_seconds, erlParams.erl_activation_period_seconds);
+      assert.equal(result.erl_quota, erlParams.erl_quota);
+      assert.equal(result.erl_quota_interval, erlParams.erl_quota_interval);
+      assert.equal(result.erl_is_active_key, erlParams.erl_is_active_key);
+      assert.equal(result.erl_quota_key, erlParams.erl_quota_key);
+      assert.equal(result.erl_configured_for_bucket, true);
     });
```
Comment (LeweyM) [inline test/utils.tests.js:242]: We could split this to try to clarify the expectations
```suggestion
       describe('when erlParams and bucketKeyConfig contain valid ERL values', () => {
      const erlParams = {
        erl_is_active_key: 'erl_is_active_key',
        erl_quota_key: 'erl_quota_key',
        erl_activation_period_seconds: 900,
        erl_quota: 10,
        erl_quota_interval: 'quota_per_calendar_month'
      };
      const bucketKeyConfig = {
        size: 1,
        interval: 60000,
        per_interval: 1,
        ttl: 60,
        ms_per_interval: 0.000016666666666666667,
        drip_interval: 60000,
        elevated_limits: {
          size: 2,
          interval: 60000,
          per_interval: 2,
          ttl: 60,
          ms_per_interval: 0.000033333333333333335,
          drip_interval: 30000,
          erl_configured_for_bucket: true,
        }
      };
      
      const result = resolveElevatedParams(erlParams, bucketKeyConfig);
      
      it('should use the bucket details from bucketKeyConfig.elevated_limits', () => {
        assert.equal(result.ms_per_interval, bucketKeyConfig.elevated_limits.ms_per_interval);
        assert.equal(result.size, bucketKeyConfig.elevated_limits.size);
        assert.equal(result.interval, bucketKeyConfig.elevated_limits.interval);
        assert.equal(result.per_interval, bucketKeyConfig.elevated_limits.per_interval);
        assert.equal(result.ttl, bucketKeyConfig.elevated_limits.ttl);
        assert.equal(result.drip_interval, bucketKeyConfig.elevated_limits.drip_interval);
      });
      it('should use the quota, activation period and keys from erlParams', () => {
        assert.equal(result.erl_activation_period_seconds, erlParams.erl_activation_period_seconds);
        assert.equal(result.erl_quota, erlParams.erl_quota);
        assert.equal(result.erl_quota_interval, erlParams.erl_quota_interval);
        assert.equal(result.erl_is_active_key, erlParams.erl_is_active_key);
        assert.equal(result.erl_quota_key, erlParams.erl_quota_key);
        assert.equal(result.erl_configured_for_bucket, true);
      });
    })
```

File: test/validation.tests.js [hunk 2]
```diff
@@ -122,4 +122,84 @@ describe('validation', () => {
...
+        erl_activation_period_seconds: 900,
+        erl_quota_interval: 'quota_per_calendar_month',
+      };
+
+      const result = validateERLParams(erlParams);
+
+      assert.strictEqual(result.message, 'a valid quota amount per interval is required for elevated limits');
+      assert.deepEqual(result.extra, { code: 112 });
+    });
+  });
```
Comment (LeweyM) [inline test/validation.tests.js:204]:
```suggestion
  describe('validateERLParams', () => {
    let validParams;
    beforeEach(() => {
      validParams = {
        erl_is_active_key: 'erl_is_active_key',
        erl_quota_key: 'erl_quota_key',
        erl_activation_period_seconds: 900,
        erl_quota: 10,
        erl_quota_interval: 'quota_per_calendar_month'
      };
    });

    it('should return no error for appropriate keys', () => {
      const result = validateERLParams(validParams);
      assert.isUndefined(result);
    });
    describe('when providing invalid parameters', () => {
      const missingParamsCases = [
        { missingParam: 'erl_is_active_key', code: 108, message: 'erl_is_active_key is required for elevated limits' },
        { missingParam: 'erl_quota_key', code: 110, message: 'erl_quota_key is required for elevated limits' },
        { missingParam: 'erl_activation_period_seconds', code: 111, message: 'erl_activation_period_seconds is required for elevated limits' },
        { missingParam: 'erl_quota', code: 112, message: 'a valid quota amount per interval is required for elevated limits' },
        { missingParam: 'erl_quota_interval', code: 112, message: 'a valid quota amount per interval is required for elevated limits' }
      ];
      missingParamsCases.forEach(testcase => {
        it(`should return ${testcase.message} if ${testcase.missingParam} is not provided`, () => {
          const invalidParams = _.omit(validParams, testcase.missingParam); 
          const result = validateERLParams(invalidParams);

          assert.strictEqual(result.message, testcase.message);
          assert.deepEqual(result.extra, { code: testcase.code });
        });
      });
    });
  });
```

File: lib/db.js [hunk 3]
```diff
@@ -290,26 +290,20 @@ class LimitDBRedis extends EventEmitter {
...
 
   takeElevated(params, callback) {
-    const valError = validateERLParams(params.elevated_limits);
-    if (valError) {
-      return callback(valError)
+    let erlParams;
+
+    if (params.elevated_limits) {
+      erlParams = utils.getERLParams(params.elevated_limits);
+      const valError = validateERLParams(erlParams);
```
Comment (LeweyM) [inline lib/db.js:297]: I think this was commented in another PR, but I think it's odd to validate _after_ parsing. 

It's fine for me in this PR, but we can take a look at this for future general refactoring.

**Review by LeweyM — APPROVED** [review]
I think we're getting to the point where it's becoming quite complicated and we can make some nice simplifications. 

I would separate that to future PRs though, so we can go ahead with this one.

Approved with some suggestions for test improvements.

File: test/validation.tests.js (hunk 2 above)
Comment (pubalokta) [reply test/validation.tests.js]: thanks, great suggestion!

File: test/utils.tests.js (hunk 1 above)
Comment (pubalokta) [reply test/utils.tests.js]: applied. thanks!

File: test/utils.tests.js [hunk 4]
```diff
@@ -147,13 +139,106 @@ describe('utils', () => {
...
 
       assert.equal(result.erl_is_active_key, params.elevated_limits.erl_is_active_key);
       assert.equal(result.erl_quota_key, params.elevated_limits.erl_quota_key);
+      assert.equal(result.erl_activation_period_seconds, params.elevated_limits.erl_activation_period_seconds);
+      assert.equal(result.erl_quota, params.elevated_limits.quota_per_calendar_month);
+      assert.equal(result.erl_quota_interval, 'quota_per_calendar_month');
+    });
+  });
+
+  describe('resolveElevatedParams', () => {
```
[the remark this answers is not in the document]
Comment (pubalokta) [reply test/utils.tests.js:157]: changed

File: lib/db.js (hunk 3 above)
Comment (pubalokta) [reply lib/db.js:297]: yeah, we could debate if parsing should come before or after validating. The thing is that parsing needs to be done for extracting the quota which comes in the form of 'quota_per_calendar_month', so it gets translated into `quota_amount` and `quota_interval`. That's why we're validating after parsing. But we could circle back into this

**Pull request merged**